### PR TITLE
feature: abstract away application selector behavior  in preparation for alt-tick behavior

### PR DESCRIPTION
**Pull request by mattkae**
## What's new?
- Abstracted away the window cycling algorithm so that we can instead its use to be not only between applications but also _within_ applications

A follow up PR will integrate the window selection within apps.

### CODE REVIEW

File: src/miral/application_selector.cpp [hunk 1]
```diff
@@ -21,243 +21,439 @@
...
 using namespace miral;
 
-ApplicationSelector::ApplicationSelector(miral::WindowManagerTools const& in_tools) :
-    tools{in_tools}
-{}
+namespace miral
+{
+
+template <typename T>
+class WindowCyclingBehavior
```
Comment (mattkae) [inline src/miral/application_selector.cpp:28]: I don't particularly like this name, if you can think of something better

File: src/miral/application_selector.cpp [hunk 2]
```diff
@@ -21,243 +21,439 @@
...
+{
+public:
+    WindowCyclingBehavior(
+        std::function<bool(T const&, T const&)> comparison_func,
+        std::function<bool(T const&)> can_select,
+        std::function<void(T const&, T const&, T const&)> on_raised,
+        std::function<void(T const&)> on_cancelled,
+        std::function<void(T&)> reset
```
Comment (AlanGriffiths) [inline src/miral/application_selector.cpp:36]: OK, so this is what you were asking about? Declaring these as virtual functions vs providing them as functors?

I think what you actually want here is a traits class:

```c++
template<typename T> WindowCyclingTraits {};

struct WindowCyclingTraits<WeakApplication>
{
    static auto equals(WeakApplication const&, WeakApplication const&) => bool;
    static auto selectable(WeakApplication const&) => bool;
    ...
};
```

Then `WindowCyclingBehavior` can `using Traits = WindowCyclingTraits<T>` and `Traits::equals(...)`.

File: src/miral/application_selector.cpp (hunk 1 above)
Comment (mattkae) [reply src/miral/application_selector.cpp:28]: The class handles  ordering a stack of `T` items such that they are always in the correct Alt + Tab order that we want

File: src/miral/application_selector.cpp (hunk 2 above)
Comment (mattkae) [reply src/miral/application_selector.cpp:36]: That makes sense to me, but do the traits need to be `static`? The problem being that functions like `can_select` depend on `WindowManagerTools`, which is passed via a constructor. Would a non-static class defeat the purpose of the traits pattern?

Comment (AlanGriffiths) [reply src/miral/application_selector.cpp:36]: TPM traits are normally types or static functions. And if you're passing in `WindowManagerTools` to construct a "trait" instance, you might just as well pass it to the static method.

OTOH another design option is to template on a base class (which provides these functions), not the stored type.

File: src/miral/application_selector.cpp [hunk 3]
```diff
@@ -21,243 +21,439 @@
...
+    ) : comparison_func{comparison_func},
+        can_select{can_select},
+        on_raised{on_raised},
+        on_cancelled{on_cancelled},
+        reset{reset}
+    {
+    }
```
Comment (AlanGriffiths) [inline src/miral/application_selector.cpp:43]:
```suggestion
struct ApplicationSelector::WeakApplicationSelector
{
    using value_t = std::weak_ptr<mir::scene::Session>;

    WeakApplicationSelector(ApplicationSelector* selector, WindowManagerTools tools);
    auto equals(value_t const&, value_t const&) -> bool;
    auto selectable(value_t const&) -> bool;
    void on_selected(value_t const& previously_selected, value_t const& newly_selected, value_t const& originally_selected);
    void on_cancelled(value_t const&);
    void reset(value_t&);

    WindowManagerTools tools;
};

template <typename SelectorBase>
class TabOrderSelector : SelectorBase
{
public:

    using SelectorBase::SelectorBase;
    using T = typename SelectorBase::value_t;
```

File: src/miral/application_selector.cpp [hunk 4]
```diff
@@ -21,243 +21,439 @@
...
+        get_can_select_func(tools),
+        get_on_raised_func(tools),
+        get_on_cancelled_func(tools),
+        get_reset_func(),
+        other.window_cycler->get_focus_list(),
+        other.window_cycler->get_originally_focused(),
+        other.window_cycler->get_focused(),
+        other.window_cycler->get_is_active()
+    );
 }
```
Comment (AlanGriffiths) [inline src/miral/application_selector.cpp:295]: 1. Do we actually copy `ApplicationSelector` instances?
2. `window_cycler` should be initialised, not initialised and then assigned
3. It looks like `WindowCyclingBehavior` needs a copy constructor, not `getters` for all its state

File: src/miral/application_selector.cpp [hunk 5]
```diff
@@ -21,243 +21,439 @@
...
+        get_on_raised_func(tools),
+        get_on_cancelled_func(tools),
+        get_reset_func(),
+        other.window_cycler->get_focus_list(),
+        other.window_cycler->get_originally_focused(),
+        other.window_cycler->get_focused(),
+        other.window_cycler->get_is_active()
+    );
     return *this;
 }
```
Comment (AlanGriffiths) [inline src/miral/application_selector.cpp:312]: I guess this is partly pre-existing, but the canonical, exception safe, way to write assignment is:

```suggestion
auto ApplicationSelector::operator=(ApplicationSelector const other) -> ApplicationSelector&
{
    std::swap(tools, other.tools);
    std::swap(window_cycler, other.window_cycler);
    return *this;
}
```

File: src/miral/application_selector.cpp [hunk 6]
```diff
@@ -21,243 +21,439 @@
...
-    focus_list.push_back(application);
+    window_cycler->add(application);
 }
 
 void ApplicationSelector::advise_focus_gained(WindowInfo const& window_info)
 {
     auto window = window_info.window();
     if (!window)
     {
+        mir::log_error("ApplicationSelector::advise_focus_gained: Could not find the window");
```
Comment (AlanGriffiths) [inline src/miral/application_selector.cpp:324]: Why is this an _error_? A WindowInfo with no window sounds like invariants have been broken. In which case, should we really be ignoring the problem and carrying on?

We ought to either:

1. Assume the invariant holds; or,
2. Report the program is in a bad state: `fatal_error()`

File: src/miral/application_selector.cpp (hunk 1 above)
Comment (AlanGriffiths) [reply src/miral/application_selector.cpp:28]: `TabOrderSelector`?

**Review by AlanGriffiths — CHANGES_REQUESTED** [review]

File: src/miral/application_selector.cpp (hunk 2 above)
Comment (AlanGriffiths) [reply src/miral/application_selector.cpp:36]: Also, things line `equals`/`comparison_func` could just be found by overloading.

File: src/miral/application_selector.cpp (hunk 4 above)
Comment (mattkae) [reply src/miral/application_selector.cpp:295]: I believe `MinimalWindowManager` doesn't have copies deleted, so I needed to define this. But do we actually copy the `MinimalWindowManager`...?

**Review by AlanGriffiths — CHANGES_REQUESTED** [review]

File: src/miral/application_selector.cpp [hunk 7]
```diff
@@ -21,243 +21,360 @@
...
+class TabOrderSelector : public SelectorBase
 {
-    auto window = window_info.window();
-    if (!window)
+public:
+    using T = typename SelectorBase::value_t;
+
+    template<typename ... Args>
+    TabOrderSelector(Args &&... args): SelectorBase(std::forward<Args>(args) ...)
+    {}
```
[the remark this answers is not in the document]
Comment (mattkae) [reply src/miral/application_selector.cpp:96]: Oh neat. I didn't know about this

**Comment by mattkae** [comment]
Well, I just realized that this could be much easier :sweat_smile: A better implementation is to come!

**Comment by mattkae** [comment]
This was unnecessary

**Pull request closed without merging**